src/query/api/v1/handler/prometheus/remote/read.go

```diff
@@ -220,9 +248,12 @@ func (h *PromReadHandler) read(
 				return
 			}
 
-			// Detect clients closing connections
-			handler.CloseWatcher(ctx, cancel, w, h.instrumentOpts)
-			result, err := h.engine.ExecuteProm(ctx, query, queryOpts, fetchOpts)
+			// Detect clients closing connections.
+			if cancelWatcher != nil {
+				cancelWatcher.WatchForCancel(ctx, cancel)
```

Comment:
Seems like we pepper cancel watchers in a ton of places, is there a way to basically make this middleware instead?

Just seems very messy the setup/tear down of it (even though its useful itself).

Reply:
Might add a todo for this?